### ISSUE TITLE
BUG: Save is triggered every few seconds, leading to glitches in handwriting.

### ISSUE THREAD

**Issue by matveyshishov** [zsviczian/obsidian-excalidraw-plugin#2004]
### Have you searched for existing issues (including closed ones)?

- [x] Yes, I have searched the existing issues.

### Does this bug persist in a new vault with only Excalidraw installed?

- [x] Yes, I have verified the issue persists.

### Your environment

SYSTEM INFO:
	Operating system: ios 18.0 (Apple iPad14,5)
	Obsidian version: 1.6.7 (149)
	API version: v1.6.7
	Login status: logged in
	Catalyst license: vip
	Live preview: on
	Base theme: light
	Community theme: none
	Snippets enabled: 14
	Restricted mode: off
	Plugins installed: 77
	Plugins enabled: 24
		1: File Color v1.1.0
		2: Colorful Note Borders v0.2.4
		3: Auto Link Title v1.5.4
		4: File Explorer Note Count v1.2.2
		5: Homepage v4.0.4
		6: Image Captions v1.1.3
		7: Reveal Active File Button v2.0.3
		8: Advanced URI v1.44.1
		9: Mega Plugin v1.0.1
		10: Link Exploder v0.3.0
		11: Folder notes v1.7.26
		12: Commander v0.5.1
		13: Hotkeys++ v0.2.7
		14: Callout Manager v1.1.0
		15: Tasks v7.9.0
		16: Dataview v0.5.67
		17: Opener v2.1.2
		18: File Explorer++ v1.2.1
		19: Creases v0.7.0
		20: Periodic Notes v0.0.17
		21: Daily notes opener v2.0.2
		22: Excalidraw v2.4.1
		23: Calendar v1.5.10
		24: Chronology v1.1.11

RECOMMENDATIONS:
	Custom theme and snippets: for cosmetic issues, please first try updating your theme and disabling your snippets. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community theme and snippets.
	Community plugins: for bugs, please first try updating all your plugins to latest. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community plugins.

### Describe the bug

I hope it's not something too problematic, sorry for bothering you!

Handwriting with iPad is my 99% of usage for Excalidraw (INCREDIBLY grateful for the Excalidraw plugin for that!), which is why I have had the "Interval for autosave on Mobile" set to "practically never (5 minutes)" (and put syncing on pause during drawing sessions):

```json
{  "autosave": true,
  "autosaveIntervalDesktop": 900000,
  "autosaveIntervalMobile": 300000}
  ```
  
In 2.4.0 it worked exactly as it should, in 2.4.1 (edit: 2.4.2 as well) it keeps saving every few seconds, as soon as my hand stops touching the screen, which leads to glitches between handwriting strokes.

### Steps to reproduce

1. Open an .excalidraw.md file on iPad
2. Choose pen
3. Start writing
4. Normally the diskette symbol stays red, and handwriting is smooth, but starting with 2.4.1 it keeps autosaving immediately when you your hand stops touching the screen, and handwriting experiences glitches (e.g. a complex shape stroke becomes a dot or a short line, and sometimes several strokes don't appear immediately, and then suddenly show up as several reduced dots/lines).

### Expected behavior

Expected to happily write on screen as usual, with the diskette symbol staying red, and (almost) no glitches occuring.

### Additional context

Tried using a vault with no other plugins.
All plugins have been updated.
In the same time window as Excalidraw update to 2.4.1 from 2.4.0, only Chronology plugin was updated, and no other changes were made to the vault.

### Edit:

Immediately after having reported the bug, received a notification about 2.4.2 update, updated the plugin, killed and reopened Obsidian, the issue still persists.
Also I forgot to mention that tried to set autosave to `false` in the data.json, it's being reset to `true` again, no matter if I change it with no Obsidian running, or during Obsidian open. Setting "turn off autosave only if you know what you're doing" doesn't help as well.
And in fact noticed that on Mac it's also autosaving, just the interval is longer.

And I'm not 100% certain the issue with glitches are due to autosave, because they are not exactly correlated, but it feels just like when I had autosave and sync available, and I haven't had such glitches in a long time, so I suspect it's not just the diskette icon indicator being incorrect.

**Comment by zsviczian** [comment]
I don't think these glitches are autosave related. You can double check if autosave is working as demonstrated in the video. Place the Obsidian tools panel on the screen, write on the screen and wait until the save button changes color. That is the point when autosave happens. Based on my testing Autosave intervals work as expected. 

https://github.com/user-attachments/assets/9c17e611-2cc7-4354-bb83-04e82784f805

I have this [wiki article](https://excalidraw-obsidian.online/Hobbies/Excalidraw+Blog/WIKI/06+Troubleshooting/Troubleshooting+Performance+Issues) about performance tuning you might want to look through this. BTW, you are correct, autosave contributes to glitches, though that was to some extent addressed in the 2.4.0 release. Nothing in 2.4.1, and 2.4.2 changed regarding that enhancement. 

Save will happen after the autosave period expired and if at least 2 seconds have passed since your last freedraw stroke.... With your setting this should happen soonest every 5 minutes.

Save also happens when excalidraw loses focus, i.e. opening the sidebar or the command palette, or switching apps.

--------

After some testing I think I know the likely source of the issue... it is likely due to how you are holding the iPad in hand, when you touch the area outside excalidraw (e.g. the tab title bar or the tabs above) Excalidraw performs a Save assuming you are about to leave the view.

**Comment by matveyshishov** [comment]
Thanks, I have already read the wiki article!

As for the possibility of touching with my hand - it's certainly not that, as I tried writing by touching just with apple pencil, and as soon as the tip leaves the screen - it saves. Also, then it wouldn't be happening on the Mac, but it does the same thing there too.
There's indeed probably something that makes it think it's lost focus, and I thought at first that it was some other plugin stealing focus or smth, but then I cloned my vault, removed all plugins, and yet the issue persisted. I tried to look with developer's tools, what's going on, but couldn't figure out how to have a breakpoint or watchdog logs or smth..

**Comment by zsviczian** [comment]
I assume you restarted Obsidian after turning off all other plugins...

I wonder if this is an iOS 18 thing? thought I highly doubt. I tested on an iOS 17.5 device. But then again, if it happens on your laptop as well, then unlikely.  At any rate I can confirm it is working as expected on all my devices (MacBook, Windows Desktop, Windows Laptop, iPad iOS 17, iPad iOS 15, Android 14 (S23 Ultra)).

Instead of cloning the Vault I'd recommend creating an empty Vault and installing only Excalidraw and testing that way.

**Comment by philippstappert** [comment]
I have the same issue, though i'm not really sure whether it is related to autosave.

I have an new, empty vault and no plugins other than excalidraw installed. Very strange 🤔

**Comment by zsviczian** [comment]
@philippstappert, please run Obsidian Command Palette > Show Debug Info and share the results here. Maybe we start to see a pattern emerge.

**Comment by matveyshishov** [comment]
Alright, I think I may have stumbled on a correlation which may be a causation:
* I'm pretty sure it's related to Obsidian sync (yet it seems to be unrelated to having sync on pause or not)
* it may (I am not sure) require the .excalidraw.md file to be opened on several devices

**TL;DR spoiler: The worst case scenario suspected? Some breaking changes on iPad OS 18 could've rendered the watchdog Excalidraw uses under the hood useless, as it's triggered immediately most of the time (not always, though).**

# Devices

_(see debug info below)_
* iPad
* iPad Mini
* Mac

# Procedure

I created today morning a simple vault from scratch.
Then I tried 3 devices.
It worked as expected!
Strangely enough, even a note from an old vault behaved well initially in the day, while having sync on pause (it's an instinct for me now to put it on pause whenever I open Excalidraw, haha).

Then I enabled sync on it.
Now I had 3 devices in front of me, and started drawing in a note on one device, while looking at other screens (which had no notes open at the moment), and repeated it for each device, with sync being on/off, 6 tests in total.
On Mac and iPad mini initially no problems were observed.
On iPad almost immediately the behavior appeared, and it did NOT matter if the sync was on pause on that iPad.

So first I opened the note on all 3 devices, and made a couple strokes, with sync on:
* on Mac, the diskette stayed red, then sync reset it in about half a minute on all devices
* on iPad mini, the diskette stayed red, then sync reset it in about half a minute on all devices
* on iPad, the diskette was reset almost immediately, then in about half a minute the other devices had it reset
_(BTW, syncing worked beautifully, I didn't try to do nasty stuff, of course, no CRDT level shenanigans, but drawing on one shows on the other devices consistently)_

OK, second test was that I put sync on pause, and repeated the experiment (not supposedly the devices can't communicate):
* on Mac, the diskette stayed red, then in half a minute reset
* on iPad mini, the diskette stayed red **for the full 5 minutes!**
* on iPad, the diskette was reset almost immediately
_(note that NO syncing happened on the notes, i.e. I didn't see strokes made on one device appearing on other devices)_

# Musings

If anything, it looks like there's something special about file watching on the combination of iPad OS 18 and Obsidian. That I started observing that behavior while no relevant changes were made in Excalidraw plugin, nor version upgrades happened in Obsidian, hints that it's probably related to iPad OS, unfortunately, I don't remember if it was updated in the time window.

The aggressive diskette reset on iPad, slower on Mac, and anaemic on iPad mini hints that it may be pertinent to considerations of energy efficiency leading to liveness/status checks frequency in the OS/device in question, iPad mini being the oldest, weakest and least capable of the bunch.

So guess what, I disabled the sync plugin on the iPad, and yeah, look and behold, it's still a naughty boi, resetting the diskette at once. You'd think it rests the case, it's just the OS file watcher problem? **No!** Remember that it didn't reset the diskette every second when I just created the note on that iPad locally earlier in the day! And sometimes it doesn't reset it even in an old fat vault with tons of plugins.

Unfortunately, I'm very constrained in terms of available time right now, otherwise I would've already done due diligence tests with network on/off and packet sniffing to see if despite claiming to be on pause Obsidian sync still captures/sends state (even if not updating the notes), but I **VERY** strongly suspect that I would've discovered that to be not the case, because (it's very simplistic, may be wrong):
* Obsidian is of the same very version on both iPads.
* the OS is different


# Debug info

## iPad mini

SYSTEM INFO:
	Operating system: ios 17.5.1 (Apple iPad14,1)
	Obsidian version: 1.6.7 (149)
	API version: v1.6.7
	Login status: logged in
	Catalyst license: vip
	Live preview: on
	Base theme: adapt to system
	Community theme: none
	Snippets enabled: 0
	Restricted mode: off
	Plugins installed: 1
	Plugins enabled: 1
		1: Excalidraw v2.4.2

RECOMMENDATIONS:
	Community plugins: for bugs, please first try updating all your plugins to latest. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community plugins.

## iPad

SYSTEM INFO:
	Operating system: ios 18.0 (Apple iPad14,5)
	Obsidian version: 1.6.7 (149)
	API version: v1.6.7
	Login status: logged in
	Catalyst license: vip
	Live preview: on
	Base theme: adapt to system
	Community theme: none
	Snippets enabled: 0
	Restricted mode: off
	Plugins installed: 1
	Plugins enabled: 1
		1: Excalidraw v2.4.2

RECOMMENDATIONS:
	Community plugins: for bugs, please first try updating all your plugins to latest. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community plugins.


## Mac

SYSTEM INFO:
	Obsidian version: v1.6.7
	Installer version: v1.4.13
	Operating system: Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000 23.5.0
	Login status: logged in
	Catalyst license: vip
	Insider build toggle: off
	Live preview: on
	Base theme: adapt to system
	Community theme: none
	Snippets enabled: 0
	Restricted mode: off
	Plugins installed: 1
	Plugins enabled: 1
		1: Excalidraw v2.4.2

RECOMMENDATIONS:
	Community plugins: for bugs, please first try updating all your plugins to latest. If still not fixed, please try to make the issue happen in the Sandbox Vault or disable community plugins.

**Comment by zsviczian** [comment]
Interesting findings. I will Experiment a little myself. I have six devices (windows desktop, macbook, iPad, Android phone, windows laptop). Obsidian Sync is always enabled. I've built some logic into Excalidraw to be intelligent about parallel changes on two devices. My scenario was to work on the desktop and use iPad for occasional hand-drawn scribble and to allow almost real time collaboration via shared Vaults over Obsidian Sync. This feature works pretty well (as per my last testing, albeit several months ago). Since then I have a drawing tablet connected to my desktop, so the need for this feature from my end is gone.

I'll play with this a bit tomorrow to see if I can somehow reproduce the issue. Until I can I won't be able to fix it. If it is an iPad 18 thing, then you are in bad luck, I don't have an iOS 18 eligible device and I don't plan on getting one... plus Apple makes it unreasonably hard (in case of Obsidian impossible) to properly debug Excalidraw on iOS, so even if I had a device working out what might be happening is questionable.

File changes could be a source of change trigger, but if sync is off, then that should not happen, unless iCloud plays into the picture somehow.

**Comment by zsviczian** [comment]
OK. I see iOS18 is only Beta at the moment. I have iOS18 Beta available for install, however I will not install it as I need my iPad for work and will not risk anything breaking. I recommend downgrading your iPad to a stable version of iOS and testing again. 

In terms of the mix of devices, it would be good to performa  clean test with only your iPad Mini and Mac.
I now tested (with sync enabled all the time between my iPad 17.6.1 and Macbook everything runs as expected.

The other difference I spotted is that I have Obsidian 1.7.1. on both my Mac and IPad. I would imagine this will be soon released to the public. Maybe this will resolve the issues - though I don't have high hopes based on my memory of the changelog for 1.7.0 and 1.7.0

**Comment by matveyshishov** [comment]
Thank you very much for taking a look!

Frankly, I don't think it's worth your effort trying to solve it right now. Who knows what kind of tracing is enabled in iPad OS 18, being a beta? It may be full of materialized telemetry triggering all kinds of behavior which would not be observed in a release version.

I've just switched to my mini for taking notes, just a barely noticeable speed bump.
It's very possible even that the number of affected people is in single digits, and the number of those noticing is even smaller. I just pay attention to these things. And even if I need a bigger screen for writing/drawing, I have other drawing apps syncing PDFs and renders into Obsidian-visible folders, I just keep trying to cut the number of scripting scaffolding and app switching to the minimum, and thus really appreciate being able to use Excalidraw because of its integration with Obsidian.

So, no worries, and thanks a lot again for your great work!

**Comment by Abahabahabach** [comment]
Actually, I encountered the same issue when I'm using graphics tablet to draw on the mac. The glitches occurred every time the stylus tip leaves the tablet's surface, coinciding with the autosave triggering.I've created an empty repository to test this, and the issues persist.However, my work depends on the graphics tablet, so my solution was to install the version 2.3.0, in which the issues do not persist.

https://github.com/user-attachments/assets/8f1260c9-6ce1-4cf1-aedb-8418865362a6

**Comment by zsviczian** [comment]
There are two event handlers that I added with 2.4.0. 
- Save on window blur: hen you switch to another app on your mobile, or you click to another app on your desktop). This prevents dataloss when iOS or Android recycles Obsidian or when the user forgets about half finished work and just shuts down their desktop, etc.
- Save on content area mouse leave: when you leave the Excalidraw working area, i.e. move to another tab in Obsidian. The assumption here is you are moving to another tab and who knows if you'll forget to come back or you may want to see your updated drawing in the document.

Because we have two new event handlers we have four options
- Excalidraw 2.4.0, 2.4.1 and 2.4.2 has both event handlers. 
- [main.js-noMouseLeaveANDnoOnBlur.zip](https://github.com/user-attachments/files/16920817/main.js-noMouseLeaveANDnoOnBlur.zip) has none of the two event handlers
- [main.js-noMouseLeaveEvent.zip](https://github.com/user-attachments/files/16920818/main.js-noMouseLeaveEvent.zip) has the onBlur, but not the mouse leave handler
- [main.js-noOnBlurEvent.zip](https://github.com/user-attachments/files/16920820/main.js-noOnBlurEvent.zip)has the mouse leave but not the onblur.

I ask for the following help.
Unzip the 3 files one by one. Copy the main.js to `.obsidian/plugins/obsidian-excalidraw-plugin` folder and restart Obsidian, then test.
It is very easy to get confused with 3 main.js-es all called the same, so I recommend unzipping one-by-one, moving the file to plugin folder, restarting, testing, then writing down the result, then doing the next.

2.4.0 is far superior from a pen handling perspective than 2.3.0 so I hope this fixes the issue for you @Abahabahabach

**Comment by zsviczian** [comment]
I just tested these on my Huion tablet. Indeed, when I lift the pen save triggers.
I can imagine that iOS 18 implemented sensitivity to detect pen proximity - so could be the root cause.

This version fixed the issue for me on my Huion:  [main.js-noMouseLeaveEvent.zip](https://github.com/user-attachments/files/16920818/main.js-noMouseLeaveEvent.zip)

**Comment by Abahabahabach** [comment]
Thank you so much for the detailed instructions and support. I am delighted to report that my issue has been completely resolved!

But I am wondering if it would be necessary to add an option in the Excalidraw Settings to enable or disable the MouseLeaveEvent?

**Comment by zsviczian** [comment]
Yeah - I will remove mouse leave event from the code.
The dilemma is to release 2.4.3 updated or not. People get concerned with too many updates, even if a tiny such as removing an event listener. I will release an update via BRAT if people want to install it, and leave this issue open, hoping that if this is causing pain for someone they find it and can fix it for themselves before the October release.

Let's also wait for @matveyshishov's confirmation before we celebrate success! But it is looking promising.

**Comment by philippstappert** [comment]
Yeah, i extracted the NoMouseLeaveEvent.zip into the folder and now it works on my machine too.

**Comment by philippstappert** [comment]
> @philippstappert, please run Obsidian Command Palette > Show Debug Info and share the results here. Maybe we start to see a pattern emerge.

Just for record, but it is fixed for me now:

```
SYSTEM INFO:
	Obsidian version: v1.6.7
	Installer version: v1.6.5
	Operating system: Windows 11 Home 10.0.22631
	Login status: logged in
	Catalyst license: none
	Insider build toggle: off
	Live preview: on
	Base theme: adapt to system
	Community theme: none
	Snippets enabled: 0
	Restricted mode: off
	Plugins installed: 1
	Plugins enabled: 1
		1: Excalidraw v2.4.2
```

**Comment by matveyshishov** [comment]
Oh, it makes perfect sense. It didn't occur to me that M1 has a different pencil hardware.

> * [main.js-noMouseLeaveANDnoOnBlur.zip](https://github.com/user-attachments/files/16920817/main.js-noMouseLeaveANDnoOnBlur.zip) has none of the two event handlers
Yep, the diskette stays red.

> * [main.js-noMouseLeaveEvent.zip](https://github.com/user-attachments/files/16920818/main.js-noMouseLeaveEvent.zip) has the onBlur, but not the mouse leave handler
Same, works as expected.

> * [main.js-noOnBlurEvent.zip](https://github.com/user-attachments/files/16920820/main.js-noOnBlurEvent.zip)has the mouse leave but not the onblur.
The diskette is reset the moment the pencil is removed far enough from the screen.

OK, confirmed it 100% by going on that M1 iPad into **Settings -> Apple Pencil** and disabling there **Hover Effects**, the diskette symbol stays red now no matter which version of main.js is used, and it works on the stock version of 2.4.2.

Thank you so very much, Zsolt!

**Comment by zsviczian** [comment]
The fix is also available if you install 2.4.3-rc-1 using BRAT. More on BRAT and beta releases here: https://youtu.be/2poSS-Z91lY

**Comment by matveyshishov** [comment]
Thank you very much again!

I'll keep main.js for the time being, it's easier ;)

**Comment by zsviczian** [comment]
resolved in 2.4.3
